hdf5-rs/src/hl/container.rs

```diff
@@ -56,22 +56,18 @@ impl<'a> Reader<'a> {
     }
 
     /// Reads a slice of an n-dimensional array.
-    /// If the array has a fixed number of dimensions, it must match the dimensionality of
-    /// dataset. Use the multi-dimensional slice macro `s![]` from `ndarray` to conveniently create
+    /// If the dimensionality `D` has a fixed number of dimensions, it must match the dimensionality of
+    /// the slice, after singleton dimensions are dropped. 
+    /// Use the multi-dimensional slice macro `s![]` from `ndarray` to conveniently create
     /// a multidimensional slice.
-    pub fn read_slice<T, S, D>(&self, slice: &SliceInfo<S, D>) -> Result<Array<T, D>>
+    pub fn read_slice<T, D>(&self, slice: &AsRef<[SliceOrIndex]>) -> Result<Array<T, D>>
```

Comment:
@Enet4 happy to have more eyes!

> Is this an attempt to reduce the number of type parameters, as suggested in #26? I honestly don't feel that it's worth introducing trait objects for only that purpose. We can make the turbo fish disappear by annotating the destination:
> 
> ```rust
> let x: Array<Pixel, _> = container.read_slice(s![.., .., 0])?;
> ```
Yes, the motivation was to simplify turbofish calling of this method. I agree that putting annotations on the destination let's you avoid turbofish, when you're putting the result into a local variable. Are there cases where it's not convenient to call this method this way? @aldanor does that reduce your desire to remove `S`? See the downside of removing the `S` below.

> It's disappointing to see a full slice info value turn into a trait object here. I understand that we don't know the size of the container in compile time, but what is stopping us here from accepting an arbitrary `SliceInfo<T, D>`?
> At this point we do not even need separate methods for retrieving 1D arrays or 2D arrays. The dimensions of the array are unambiguous because the slice outlines the expected output dimensionality.
> In the current version of the code, `D` is no longer bound to the slice argument, requiring us to redundantly specify it in other cases (3+ D arrays).

Agreed -- the downside of the trait object is that the static `D` of the slice is lost, so it can't be connected to the output `D`, which makes this API less ergonomic.

I'm not sure which way to go here.

Reply:
OK - I've reverted the trait-object style API.

Re the `read_1d` methods. Here's all the ways to read a 1d slice from a 2d dataset:
```rust
    // No import, 1 placeholder
    let dsc = rdr.read_slice_1d::<usize,_>(s![..5, 5])?;

    // Must import Array1, no placeholders
    let dsc: Array1<usize> = rdr.read_slice_1d(s![..5, 5])?;   

    // No import, 2 placeholders
    let dsc = rdr.read_slice::<usize, _, _>(s![..5, 5])?;

    // Must import, no placeholders
    let dsc: Array1<usize> = rdr.read_slice(s![..5, 5])?;
```

Pros:
- one fewer placeholder in turbofish syntax
- consistency with non-sliced methods

Cons:
- No improvement over `read_slice` in non-turbofish syntax. (Non-turbofish is probably what we'd put in examples)

@aldanor I like the `ds.slice(s![...]).read()` proposal because it avoid placeholders _and_ doesn't requiring importing `Array` types, which is an improvement over the options above. The downside is that it's confusing -- the only thing you can do with the output of `ds.slice()` is call `.read()`, and there's no reason to reuse a 'sliced reader', so why is it split into 2 calls? We could help things a bit by putting a `#[must_use]` on `slice()`.

As long as there's some good examples in the docs, I think any of these options will be easily learnable.

Reply:
@Enet4 any thoughts?

(if we're all being unsure, we'd probably just merge this in anyway since it adds a bugfix, with a possibility of revamping the slice API later)